src/test/lib/DataTest.cpp

```diff
@@ -1857,3 +1857,302 @@ QuicTestReceiveResumeNoData(
+            break;
+        }
+        default:
+            break;
+        }
+    }
+    return Status;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+_Function_class_(QUIC_CONNECTION_CALLBACK)
+static
+QUIC_STATUS
+QUIC_API
+QuicAckDelayConnectionHandler(
+    _In_ HQUIC QuicConnection,
+    _In_opt_ void* Context,
+    _Inout_ QUIC_CONNECTION_EVENT* Event
+    )
+{
+    AckSendDelayTestContext* TestContext = (AckSendDelayTestContext*)Context;
+    if (TestContext->ServerConnection == QuicConnection) {
+        //
+        // Server side
+        //
+        switch (Event->Type) {
+        case QUIC_CONNECTION_EVENT_PEER_STREAM_STARTED:
+            MsQuic->SetCallbackHandler(
+                Event->PEER_STREAM_STARTED.Stream,
+                (void*)QuicAckDelayStreamHandler,
+                Context);
+            TestContext->ServerStream.Handle = Event->PEER_STREAM_STARTED.Stream;
+            QuicEventSet(TestContext->ServerStreamStartedEvent.Handle);
+            break;
+        default:
+            break;
+        }
+    } else {
+        if(TestContext->ClientConnection.Handle != QuicConnection) {
+            TEST_FAILURE("Client connection is wrong?! %p vs %p",
+                TestContext->ClientConnection.Handle,
+                QuicConnection);
+        }
+        //
+        // Client side
+        //
+        switch(Event->Type) {
+        case QUIC_CONNECTION_EVENT_CONNECTED:
+            QuicEventSet(TestContext->ClientConnectedEvent.Handle);
+            break;
+        default:
+            break;
+        }
+    }
+    return QUIC_STATUS_SUCCESS;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+_Function_class_(QUIC_LISTENER_CALLBACK)
+static
+QUIC_STATUS
+QUIC_API
+QuicAckDelayListenerHandler(
+    _In_ HQUIC /* QuicListener */,
+    _In_opt_ void* Context,
+    _Inout_ QUIC_LISTENER_EVENT* Event
+    )
+{
+    AckSendDelayTestContext* TestContext = (AckSendDelayTestContext*)Context;
+    switch (Event->Type) {
+        case QUIC_LISTENER_EVENT_NEW_CONNECTION:
+            TestContext->ServerConnection.Handle = Event->NEW_CONNECTION.Connection;
+            MsQuic->SetCallbackHandler(TestContext->ServerConnection.Handle, (void*) QuicAckDelayConnectionHandler, Context);
+            return MsQuic->ConnectionSetConfiguration(Event->NEW_CONNECTION.Connection, TestContext->ServerConfiguration);
+        default:
+            TEST_FAILURE(
+                "Invalid listener event! Context: 0x%p, Event: %d",
+                Context,
+                Event->Type);
+            return QUIC_STATUS_INVALID_STATE;
+    }
+}
+
+void
+QuicTestAckSendDelay(
+    )
+{
+    uint32_t TimeoutMs = 3000;
+    uint32_t AckDelayMs = 1000;
+    MsQuicRegistration Registration;
+    TEST_TRUE(Registration.IsValid());
+
+    MsQuicAlpn Alpn("MsQuicTest");
+
+    MsQuicSettings Settings{};
+    Settings.SetIdleTimeoutMs(TimeoutMs);
+    Settings.MaxAckDelayMs = AckDelayMs;
+    Settings.IsSet.MaxAckDelayMs = true;
```

Comment:
I feel like we should have a rule: If you add code to set a setting that doesn't have a helper yet, you get to add the helper. :D

Reply:
Good rule!